File: apps/api/src/config.test.ts
import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, describe, expect, it } from "vitest";

import { loadConfig } from "./config";

const tempDirs: string[] = [];

afterEach(() => {
  for (const dir of tempDirs.splice(0)) rmSync(dir, { recursive: true, force: true });
});

function secret(value: string): string {
  const dir = mkdtempSync(join(tmpdir(), "nudgeon-config-"));
  tempDirs.push(dir);
  const path = join(dir, "value");
  writeFileSync(path, `${value}\n`, { mode: 0o600 });
  return path;
}

describe("loadConfig *_FILE support", () => {
  it("loads required connection settings from secret files", () => {
    const config = loadConfig({
      DATABASE_URL_FILE: secret("postgres://nudgeon:secret@postgres:5432/nudgeon"),
      REDIS_URL_FILE: secret("redis://:secret@redis:6379"),
      CLICKHOUSE_URL_FILE: secret("http://nudgeon:secret@clickhouse:8123/nudgeon"),
      MODE: "single_tenant",
    });

    expect(config.databaseUrl).toContain("postgres://nudgeon:secret");
    expect(config.redisUrl).toContain("redis://:secret");
    expect(config.clickhouseUrl).toContain("http://nudgeon:secret");
    expect(config.mode).toBe("single_tenant");
  });

  it("rejects ambiguous inline and file values", () => {
    expect(() =>
      loadConfig({
        DATABASE_URL: "postgres://inline",
        DATABASE_URL_FILE: secret("postgres://file"),
        REDIS_URL: "redis://redis:6379",
        CLICKHOUSE_URL: "http://clickhouse:8123",
      }),
    ).toThrow("DATABASE_URL와 DATABASE_URL_FILE은 동시에 설정할 수 없습니다");
  });
});

File: apps/worker/internal/config/config_test.go
package config

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func writeSecret(t *testing.T, value string) string {
	t.Helper()
	path := filepath.Join(t.TempDir(), "value")
	if err := os.WriteFile(path, []byte(value+"\n"), 0o600); err != nil {
		t.Fatal(err)
	}
	return path
}

func TestLoadFromFiles(t *testing.T) {
	t.Setenv("DATABASE_URL", "")
	t.Setenv("REDIS_URL", "")
	t.Setenv("CLICKHOUSE_URL", "")
	t.Setenv("DATABASE_URL_FILE", writeSecret(t, "postgres://nudgeon:secret@postgres:5432/nudgeon"))
	t.Setenv("REDIS_URL_FILE", writeSecret(t, "redis://:secret@redis:6379"))
	t.Setenv("CLICKHOUSE_URL_FILE", writeSecret(t, "http://nudgeon:secret@clickhouse:8123/nudgeon"))

	cfg, err := Load("DATABASE_URL", "REDIS_URL", "CLICKHOUSE_URL")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(cfg.DatabaseURL, "nudgeon:secret") || !strings.Contains(cfg.RedisURL, ":secret") {
		t.Fatalf("secret file values were not loaded: %#v", cfg)
	}
}

func TestLoadRejectsInlineAndFile(t *testing.T) {
	t.Setenv("DATABASE_URL", "postgres://inline")
	t.Setenv("DATABASE_URL_FILE", writeSecret(t, "postgres://file"))
	_, err := Load("DATABASE_URL")
	if err == nil || !strings.Contains(err.Error(), "동시에 설정할 수 없습니다") {
		t.Fatalf("expected ambiguity error, got %v", err)
	}
}
